Complex grid toevoegendata

gisaac85 opened the pull request:
fix a bug when select AttributeType in ComplexGrid1 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Form fields now reliably write updates back to models with improved type conversion (including enums), null-safety and conversion error handling to prevent invalid updates.
  * Attribute type selection no longer auto-selects a default, preventing unexpected changes when no selection is provided.

* **Refactor**
  * Attribute type picker simplified to use the parent-provided value as the single source of truth, removing internal selection state.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
fix: Refactor AttributeTypePicker and enhance form binding

Updated `AttributeTypePicker.razor` to improve type safety by using the actual `AttributeTypeModel` type for the `RadzenDropDown` component. Simplified state management by removing internal state and adjusting the `OnSelectedItemChanged` method to directly set the `Value` based on the selected item's ID.

Enhanced `SetValue` method in `InnovativeForm.razor.cs` to ensure that changes to form values are reflected in the underlying model, improving data binding and encapsulation.

Commit:
Merge branch 'develop' into 71_complexGrid_toevoegendata

Commit:
Refactor RadzenDropDown in AttributeTypePicker.razor

Updated TItem, TextProperty, and ValueProperty to use
nameof for improved maintainability and clarity.

Commit:
fix: Refactor AttributeTypePicker and enhance form value handling

Updated AttributeTypePicker to use nullable int for Change event, improving type safety. Simplified OnSelectedItemChanged method to directly use the value parameter. Enhanced SetValue method in InnovativeForm to include type conversion logic, improving error handling and ensuring correct property assignments.

Commit:
fix: Ensure Value is updated consistently in AttributeTypePicker

Updated the logic in `OnSelectedItemChanged` to always assign the `Value` property to `selectedType`, even when it is null. This change prevents inconsistencies by ensuring that `Value` is updated before invoking `OnValueChanged()`.

Commit:
fix: Update RadzenDropDown and improve type conversion logic

Refactor `AttributeTypePicker.razor` to use actual type
for `TItem` and add `TValue` for better type safety.

In `InnovativeForm.razor.cs`, enhance type conversion
handling by determining `targetType` outside of checks,
updating exception handling, and preventing null assignments
to non-nullable value types by using default instances.

Commit:
fix: Refactor value handling in AttributeTypePicker and InnovativeForm

- Update `OnSelectedItemChanged` to set `Value` only if `selectedType` is not null.
- Simplify property value setting in `InnovativeForm` by using `SetValue` method.
- Enhance type conversion logic to prefer `TypeConverter`, improving robustness and error handling.

Commit:
fix: Refactor InnovativeForm with and error handling

Added using directives for logging and localization. Removed duplicate using directive for System.ComponentModel. Updated error handling to store converted values in formValues, ensuring UI state consistency with the model during conversion errors.

Enhance InnovativeForm with improved value handling

Added necessary using directives and removed redundancy.
Updated form state management by assigning converted values
to formValues and notifying property changes for better
UI responsiveness.

peereflits commented on ExampleApp/Components/AttributeTypePicker.razor.cs:
```diff
@@ -16,22 +15,16 @@ protected override async Task OnInitializedAsync()
 
     protected override void OnParametersSet()
```
Remove method while body is empty

Commit:
refactor: remove unused method

peereflits commented on Src/Innovative.Blazor.Components/Components/Form/InnovativeForm.razor.cs:
```diff
@@ -403,6 +399,79 @@ private string GetStringValue(string propertyName)
+                    else
+                    {
+                        var converter = TypeDescriptor.GetConverter(targetType);
+                        if (converter != null)
```
converter never is null. Remove if

peereflits commented on Src/Innovative.Blazor.Components/Components/Form/InnovativeForm.razor.cs:
```diff
@@ -403,6 +399,79 @@ private string GetStringValue(string propertyName)
+            object? convertedValue = value;
+            var targetType = Nullable.GetUnderlyingType(propertyType) ?? propertyType;
+
+            if (value != null && !targetType.IsInstanceOfType(value))
```
`!targetType.IsInstanceOfType(convertedValue)`

peereflits commented on Src/Innovative.Blazor.Components/Components/Form/InnovativeForm.razor.cs:
```diff
@@ -403,6 +399,79 @@ private string GetStringValue(string propertyName)
+                     or NotSupportedException)
+                {
+                    // Revert the form value to keep UI state consistent with the model
+                    formValues[propertyName] = prop.GetValue(Model);
```
this causes a side effect; apply CQS

peereflits commented on Src/Innovative.Blazor.Components/Components/Form/InnovativeForm.razor.cs:
```diff
@@ -403,6 +399,79 @@ private string GetStringValue(string propertyName)
+
+            if (value != null && !targetType.IsInstanceOfType(value))
+            {
+                try
```
Method is getting to big (and has side effects). Refactor into smaller methods.
https://csharpcodingguidelines.com/member-design-guidelines/#AV1115
https://csharpcodingguidelines.com/maintainability-guidelines/#AV1547
https://csharpcodingguidelines.com/maintainability-guidelines/#AV1500

Commit:
refactor: Refactor form handling in InnovativeForm<TModel>

Moved property value setting from SetValue to OnFormSubmit.
Simplified SetValue by removing complex conversion logic and
error handling, focusing on updating formValues only.
OnFormSubmit now directly sets model properties using reflection.
